src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt

```
@@ -0,0 +1,570 @@
+                throw EncodeException("Can not encode field with name '${fieldStructure.name}'. Message must contains field '${fieldStructure.name}")
+            else
+                return
+        }
+
+        if (fieldStructure.isCollection) {
+            if (!value.hasListValue()) {
+                throw EncodeException("Can not encode field with name '${fieldStructure.name}'. Field must contains list value")
+            }
+
+            value.listValue.valuesList.forEach {
+                encodeValue(document, node, key, fieldStructure, it)
+            }
+        } else {
+            encodeValue(document, node, key, fieldStructure, value)
+        }
+    }
+
+    protected fun encodeValue(
+        document: Document,
+        node: Node,
+        key: String,
+        fieldStructure: IFieldStructure,
+        value: Value
+    ) {
+        if (fieldStructure.isComplex) {
+            if (!value.hasMessageValue() || fieldStructure !is IMessageStructure) {
+                throw EncodeException("Can not encode field with name '${fieldStructure.name}'. Field must contains message value")
+            }
+
+            encodeMessage(document, node, value.messageValue, fieldStructure)
+        } else {
+            node.appendChild(document.createElement(key)).setText(value.simpleValue, document)
+        }
+    }
+
+
+    override fun decode(messageGroup: MessageGroup): MessageGroup {
+        val messages = messageGroup.messagesList
+        if (!messages.any { it.hasRawMessage() }) {
+            return messageGroup
+        }
+
+
+        return MessageGroup.newBuilder().addAllMessages(
+            messages.flatMap { input ->
+                if (input.hasRawMessage())
+                    runCatching {
+                        decodeOne(input.rawMessage).map { AnyMessage.newBuilder().setMessage(it).build() }
+                    }
+                        .onFailure { LOGGER.debug("Can not decode message = ${input.rawMessage.toJson()}", it) }
```

a codec shouldn't hide an exception during decode, please wrap the exception to IllegalStateException, add more details to message and throw it